### DISCUSSION TITLE
Any feasible Java-free solution for Signal – in C/C++, Lisp or Ocaml?

### DISCUSSION THREAD

**Discussion by barracuda156** [AsamK/signal-cli#1360] [Q&A]
Is there any known non-Java implementation using some properly portable language, ideally C/C++? But Lisp, Ocaml or ML may work as well.

**Comment by barracuda156** [comment]
@AsamK By the way, what is the minimal Java version needed? Will openjdk8 do, headless? If yes, that may work.

**Reply by AsamK** [reply]
Minimum required java version is 21, as mentioned in the README.
There's a signal library implemented in Rust, that might work for you: https://github.com/whisperfish/presage

**Reply by barracuda156** [reply]
To be honest it is probably easier for me to fix jdk21 than to fix rust, at least the first is likely trivial technically, just ridiculous in terms of time needed, due to jdk’s abominable bootstrap process: I need to fix and build probably every jdk from 9 to 21. With rust, it totally depends on gcc progress on it. Once gccrs works, it might be possible to add support for my platform (or maybe upstream will add it anyway).
